File: src-tauri/src/agent.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// One configured AI agent. Everything except the API key is persisted in
/// SQLite; the API key itself lives in the OS credential manager, keyed by
/// `id` (see [`save_api_key`] / [`delete_api_key`] below).
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Agent {
    pub id: String,
    pub name: String,
    /// "openai" | "anthropic" | "deepseek" | ... kept as a free-form string
    /// so new providers can be added later without a schema migration.
    pub provider: String,
    pub model: String,
    pub system_prompt: Option<String>,
    pub temperature: f64,
    pub color: Option<String>,
    /// Unix timestamp (seconds since epoch).
    pub created_at: i64,
}

/// Service name under which all agent API keys are grouped in the OS
/// credential manager (Windows Credential Manager / macOS Keychain / ...).
const SERVICE_NAME: &str = "multi-agent-group-chat";

/// Store (or overwrite) the API key for the given agent id.
pub fn save_api_key(agent_id: &str, api_key: &str) -> Result<(), String> {
    let entry = keyring::Entry::new(SERVICE_NAME, agent_id).map_err(|e| e.to_string())?;
    entry.set_password(api_key).map_err(|e| e.to_string())
}

/// Remove the stored API key for the given agent id, if any.
/// Not having an entry to delete is not treated as an error.
pub fn delete_api_key(agent_id: &str) -> Result<(), String> {
    let entry = keyring::Entry::new(SERVICE_NAME, agent_id).map_err(|e| e.to_string())?;
    match entry.delete_credential() {
        Ok(()) => Ok(()),
        Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}

File: src/types/agent.ts
/** Keep in sync with the `Agent` struct in `src-tauri/src/agent.rs`. */
export interface Agent {
  id: string;
  name: string;
  provider: string;
  model: string;
  system_prompt: string | null;
  temperature: number;
  color: string | null;
  /** Unix timestamp (seconds since epoch). */
  created_at: number;
}

export const PROVIDERS = ["openai", "anthropic", "deepseek", "qwen"] as const;
export type Provider = (typeof PROVIDERS)[number];

/** Shape of the form while the user is editing it (before it becomes an `Agent`). */
export interface AgentFormValues {
  id: string;
  name: string;
  provider: string;
  model: string;
  system_prompt: string;
  temperature: number;
  color: string;
  api_key: string;
}

export const DEFAULT_COLOR = "#6366f1";

export function emptyFormValues(): AgentFormValues {
  return {
    id: "",
    name: "",
    provider: PROVIDERS[0],
    model: "",
    system_prompt: "",
    temperature: 0.7,
    color: DEFAULT_COLOR,
    api_key: "",
  };
}

export function agentToFormValues(agent: Agent): AgentFormValues {
  return {
    id: agent.id,
    name: agent.name,
    provider: agent.provider,
    model: agent.model,
    system_prompt: agent.system_prompt ?? "",
    temperature: agent.temperature,
    color: agent.color ?? DEFAULT_COLOR,
    api_key: "",
  };
}
